INTERN-2: Peft refactor

Martakor1 opened the pull request:
Made hook refactor. Now hooks creates without parameters, or with them - depends on hook child class. Parameters of hook comes from BaseCompressionModel children classes. All linkage with trainer in the trainer link_to_trainer (BaseNeuralModel).
Also there is no phase "finetune" in BasePruner, and new parameter "prune_each" in config, that allows to control pruning during training flow. There is no "ft_params" - useless.
Also lots of minor refactoring.

Martakor1 committed:
Очевидный рефакторинг + gitignore

Martakor1 committed:
Minor test refactor and finetune_config depricating in tests

Martakor1 committed:
type annotations and small fixes (not about hooks)

Martakor1 committed:
Changed hooks v1! Now they constructing in childrens of BaseCompressionModel and appending in BaseNeuralModel. Also new link_to_trainer() method. And huge refactor of BasePruner also: ft_params are deprecated, finetune stage no more exists.
BaseNM constructor changed

leostre commented on fedcore/models/network_impl/base_nn_model.py:
```diff
@@ -64,27 +65,20 @@ def __init__(self, params: Optional[OperationParameters] = None, additional_hook
         self.learning_rate = self.params.get("learning_rate", 0.001)
-        self.model = self.params.get("model", None)
-        self._init_custom_criterions(
-            self.params.get("custom_criterions", {}))  # let it be dict[name : coef], let nodes add it to trainer
```
либо убран ценный  комментарий, либо теперь не хватает пояснения, как задавать члены лоссов

Martakor1 replied:
Добавлю тогда в NeuralModelConfigTemplate, мне кажется в месте объявления этого параметра комментарий будет смотреться лучше.

leostre commented on fedcore/algorithm/pruning/hooks.py:
```diff
@@ -149,48 +122,51 @@ def prune_after_reg(self, pruner, pruning_iter):
-        self.prune_after_reg(pruner, pruning_iter)
+    def action(self, epoch, kws):
+        pruning_iter = self.pruning_iterations
+        val_dataloader = kws['val_loader']
```
я упоминал тогда вечером, что использование валидационного датасета при прунинге приводит к data leakage и необходимо выделение некоторого подмножества из трейна

leostre reviewed (approved):
Слушай, в целом, адекватно, тесты тоже достаточно логично и явно выглядят, раз проходят, видимо, хорошо. Единственное, меня смущает логика link_to_trainer и hooks  extesion относительно модели BaseNeural. Было бы славно убедиться, что вызов например, с prune_each & save_each приводит к инстанцированию как ZeroShotPruner, так и Saver

Martakor1 committed:
Refactor: BaseQuantizer and hooks.
Also add @abstractmethod fot _init_trainer_... in BaseCompressionModel
Changed some other hooks with default epoch calc
Removed qat_params
Changed tests a little

Martakor1 commented on fedcore/algorithm/pruning/hooks.py:
```diff
@@ -1,144 +1,117 @@
 class ZeroShotPruner(BaseHook):
-    _SUMMON_KEY = 'pruning'
+    _SUMMON_KEY = 'prune_each'
     _hook_place = 50
```
Поля pruning я не наблюдал и оно не работало

Martakor1 commented on fedcore/models/network_impl/base_nn_model.py:
```diff
@@ -52,9 +51,11 @@ class BaseNeuralModel(torch.nn.Module):
-    def __init__(self, params: Optional[OperationParameters] = None, additional_hooks=None):
+    BASE_HOOKS : list[type['BaseHook']] = LOGGING_HOOKS + MODEL_LEARNING_HOOKS
+
+    def __init__(self, model: torch.nn.Module, params: dict = {}, additional_hooks: Sequence['BaseHook'] = []):
```
Почему логичнее? Кажется, BaseNeuralModel нет смысла существовать без какой-то внутренней имплементации. Модель - неотъемлемая часть этого класса.

leostre replied:
это да, но часто у нас может быть создан пайплайн, а вот модель появляться только на стадии fit, потому что является частью входных данных.
Подобная логика происходит со структурой данных в sklearn, например

Martakor1 replied:
Может бы речь идёт про LowRank, BasePruner и так далее? У них осталась возможность создаваться без model. И действительно в них модель приходит на .fit(). Но для BaseNeuralModel я посмотрел - везде он инициализируется уже когда есть model, так как этот класс обычно выполняет функцию внутреннего trainer - то есть обёртки над моделью. Пока не вижу прямого использования этого класса без model.
Но если может появляться BaseNeuralModel где-то в пайплайне как самостоятельная модель, то окей, переделаю. Правда нет особо в этом смысле, потому что что она будет делать тогда, только базовые хуки использовать? То есть просто обучение пустое без ничего?

Кстати, пока что там итак стоит в BaseNM.fit() строка self.model = input_data.target if self.model is None else self.model. Мб просто пометить, что параметр это Optional[] ?

Martakor1 commented on fedcore/algorithm/low_rank/svd_tools.py:
```diff
@@ -17,11 +17,11 @@
-def decompose_module(
+def decompose_module_in_place(
     model: Module,
     decomposing_mode: Optional[str] = True,
```
Абсолютно не понятно, как может Optional[str] = True. Отслеживал по коду, сначала была строка, потом стал бул. Не ясна семантика поля.

Martakor1 replied:
Сделал

Martakor1 commented on fedcore/algorithm/pruning/hooks.py:
```diff
@@ -342,8 +314,19 @@ def eval_entropy():
+        callback = PrunerWithGrad
+    elif zeroshot_pruner:
+        callback = ZeroShotPruner
+    return callback
```
Так и не понял, что тут хотели. Очень странно отбираются прунеры. Есть пересечения в названиях и остальном. Кажется, что поломано.

Martakor1 commented on fedcore/models/network_impl/hooks.py:
Evaluator: inputs = tuple(inputs_.to(self.device) for inputs_ in inputs if hasattr(inputs_, 'to')) - разве могут быть не torch.Tensor без to?

У Freezer поле criterion означает не совсем тот criterion, который loss, может есть другое название, более подходящее?

Martakor1 commented:
В целом я оставил все вопросы, которые накопил по коду.

Martakor1 commented on fedcore/algorithm/pruning/hooks.py:
```diff
@@ -1,144 +1,112 @@
-            loss = self.criterion_for_grad(out, target)
+            loss = self.criterion_for_pruner(out, target)
         loss.backward()
         return loss
```
Я про вот это (try и except делают одно и то же)
        
```try:
            loss = self.criterion_for_grad(out, target)
        except Exception:
            loss = self.criterion_for_grad(out, target)
        loss.backward()
        return loss

Martakor1 committed:
minor refactor

Martakor1 committed:
Update PrunerWithReg to use train_dataloader for preventing data leakage

Martakor1 committed:
Change decomposing mode to two fields: make_decompose: bool and decomposing_mode: LiteraL["channel", "spatial"]

Martakor1 committed:
Change hook Freezer.criterion -> Freezer.rule

elarensb committed:
Add docstrings for low-rank, pruning and quantization modules

elarensb committed:
Add docstrings for API config templates

elarensb committed:
Add docstrings

Lopa10ko committed:
feat: add transmla, llm-trainer, peft strategies, model registry (#42)

* creating llm trainer and testing

* added rp-hosvd, r-sthosvd and r-st algorithms (+tests)

* пока в работе метрики нлп

* refactor(metrics): unify style, docstrings, add smoke tests

* Revert "added rp-hosvd, r-sthosvd and r-st algorithms (+tests)"

This reverts commit e4e90aab58fb42d36a67497a0cf491ece9d976a9.

revert decompositionse

* Added basic implementation of interfaces. Added changes to LLMTrainer and BaseNeuralModel classes. Changed type definition logic in trainer_factory

* Added trainer selection function

* Updated llm_trainer, trainer_factory unit tests

* added transmla

* Amendments made according to review. Plugins for NLP tasks added. Task set in API expanded (QA and summarization added)

* Added connection between FedCore and Transformer Trainer hooks

* refactoring according chertila1's review

* Adjustments made based on review

* transmla refactoring according to "good code" and plan presentations + tests

* Updated config structure. Fixed llm_trainer

* rollback to right versions of functions _recreate_decomposed_*

* deleted stub functions

* deleted DefferedConversion class

* removed backward compatibility

* replaced ReassemblerFactory with simple reassembler registry

* removed TRANSMLA_PATH local variable

* removed local variables with decent stuff

* replaced "and" with "if" :)

* made BaseReassembler as specific pipeline node

* moved code with reassembling logic to other places

* fixed imports of ParentalReassembler and decomposed recrestion functions

* deleted "save_path" things, _convert_modules wrap, get_recreation_function

* moved qwen transmla demo to tmp

* removed tokenizer assert

* replaced TransMLAImporter class with _initialize_transmla() private function

* replaced dict .get() method in conversion_map with key value achieving

* refactored reassemblers dict

* added recreation functions into supported_decomposed_layers dict

* Added file for training the model using LLMTrainer. Changes have been made to ensure the model works correctly.

* made unite logic for reassemble() method

* removed AttentionReassembler, because it was too ambiguous, and it contained a lot of logic duplication

* made clean version of low_rank_opt.py
checked on test_lr.py - works correct

* made clean version of low_rank_opt.py
checked on test_lr.py - works correct

* TransMLA integration into LowRankModel node

* Added registry entity

* deleted reassembly hooks

* deleted BaseReassembler algorithm strategy

* moved reassembly directory into low_rank directory

* moved reassembly directory into low_rank directory

* Replaced deepcopy with the ModelRegistry module

* update of reassembly unit tests

* made flat-llm py-module

* WIP: trying to make flatllm reassembler

* fixed get_transmla_path() function

* Added nlp configs

* Fixed incorrect imports

* Fixed according to review

* Changed Fedot methods to replace data structure

* added auto-config + demo

* справления согласно комментариям в PR

* обавлены исправленные метрики

* Commit to save last version of files

* Fixed according to review

* Added saving of pipeline parameters

* Added LR+LLm example (doesn't work yet)

* Changed decomposed_layers file to run example file

* Adjustments have been made to run file

* Fixed bugs

* Created LowRank+LLM example. Changes have been made to low_rank_opt to run the model

* Changes according to review

* Added method overriding to match the Singleton pattern

* Added tdecomp import. Added access to decompositions via a module. Added support for the ULTG and ParallelTG optimizers.

* Refactored nlp_metrics.py: added QualityMetric compatibility

* Removed absolute links, extra import. Changed export method in main

* feat(flatllm): refactored FLAT-LLM module with adding absorption

- Add AbsorptionCompressor with Nyström method for MLP and head-wise PCA for attention
- Implement activation collection and eigenvector computation
- Add runtime forward patching for compressed attention layers
- Support GQA (Grouped Query Attention) architectures
- Physical dimension reduction: 8.6% size decrease on TinyLlama-1.1B
- Add generation speed benchmarking in test script
- Fix Q/K projections
- Compatible with Llama, Mistral, Qwen, and other Llama-based models

* Typo in folder fedcore/architecture/computational name corrected. Added use of NLP metrics

* Refactor: Rewritten using PyTorch for improved performance and consistency, replacing NumPy-based implementation.

* Fix compose mode logic, add hook support, fix embedding transpose. Trainer classes have been changed - duplicated logic has been removed, the base class has been updated

* Module restructured — ModelRegistry class has been split into logical modules. Changed ModelRegistry to ensure it complies with Singleton.

* feat(flatllm): made FlatLLM class based on Reassembler architecture (+demo example)

* Cleaned api examples

* refactored some examples and training strategy

* efactor metrics system: migrate from umy to yorch and add comprehensive tests

-  : ll metrics now use yorch tensors instead of umy

* WIP(CaRA): implementing CaRA method (status: tests working good, but it still need architecture enhancements)

* deleted accidental exps

* refactored some metrics

* Attempts to get rid of hardcoded code. Model passed as data type via kwargs, not params. Add is_llm config priority and model access properties

* Add metrics API tests and code refinements

- Add API tests for metrics with working implementation
- Minor code fixes and improvements across metrics modules
- Remove unused test file

* Commit to save minor changes

* Added thread-safe realization. Removed get_instance call

* classmethod replaced with self_registry. Device definition changed

* Replaced console output with logging

* The class call has been replaced with self._registry

* Updated logging (with constructor injection)

* fix: rm typo in logger info call

* Deleted extra directory

* Refactored ModelRegistry for correct operation

* Start of experiments with ModelRegistry

* refactor(CaRA): fix critical safety issues and improve code quality

Critical fixes:
- Replace unsafe weight mutation with F.multi_head_attention_forward()
- Fix Python 3.9+ compatibility (Union type hints)

Improvements:
- Use CPD consistently for bias adaptation via CaraAdapter
- Add validation for tensors up to 52 modes
- Remove dead code in CPDecomp.reconstruct()
- Add comprehensive docstrings (Args/Returns/Raises)
- Improve error messages with actual values
- Use torch.allclose() for reliable float comparisons

* metrics: unify tests; fix bugs

* api config changes

* repository changes wrt to torchmetrics

* ref: torchmetrics added

* fix: distillation for new metrix, placeholders

* Added first version of examples of ModelRegistry performance (will be changed)

* deleted TestFlatLLMRegistrationSafe

* fixed analyzing model's device

* removed AutoConfig logic duplication from model

* included AutoConfig check into TransMLA Reassembler

* FlatLLM example parametrization

* replace analyze_model/hardware functions with mixin classes for configuration analysis

* added support of different dataset formats

* fixed imports

* commit for safety

* Working ModelRegistry version

* Added examples. Dirty code

* api evaluate_metric changes; fedcore replacements for MetricByTask and MetricsRepository

* test: MetricFactory

* test: MetricFactory

* ruler refactoring

* metrics tests

* Fixed data saving format

* refactored PerformanceEvaluator

* Added time measurement logs to LowRank + Pretrain model example

* Fixed CheckpountManager, added accidentally removed methods

* Added changed LowRank + Pretrained model log file

* mlflow reassembly experiment

* Очевидный рефакторинг + gitignore

* Minor test refactor and finetune_config depricating in tests

* type annotations and small fixes (not about hooks)

* Changed hooks v1! Now they constructing in childrens of BaseCompressionModel and appending in BaseNeuralModel. Also new link_to_trainer() method. And huge refactor of BasePruner also: ft_params are deprecated, finetune stage no more exists.
BaseNM constructor changed

* feat: support of computational with MetricFactory

* test: computational metrics

* Fixed ModelRegistry, FedcoreDispatcher, FedcoreOptimizer to perform evolutionary optimization. Code with logs e.t.c (really dirty)

* Refactor: BaseQuantizer and hooks.
Also add @abstractmethod fot _init_trainer_... in BaseCompressionModel
Changed some other hooks with default epoch calc
Removed qat_params
Changed tests a little

* Changing the logic of saving metrics

* Changing the output in predict_for_fit

* Changed pruning and small LLM experiments for fit tests

* Changed prune_conv_model.py to to test the fit for the training strategy

* minor refactor

* Attempt to resolve problems with data input/output (DataConverter created)

* gitignore csv and drop unnecessary line

* Update PrunerWithReg to use train_dataloader for preventing data leakage

* Change decomposing mode to two fields: make_decompose: bool and decomposing_mode: LiteraL["channel", "spatial"]

* Change hook Freezer.criterion -> Freezer.rule

* temp: Inherit CompressionInpitData from InputData. Replace the use of InputData with CompressionInputData in functions. Add hardcoded solution to the multiclass-multioutput problem.

* Add docstrings for low-rank, pruning and quantization modules

* fix: fix bugs in CompressionInputData initialization, device handling (CPU/CUDA), and safe extraction of predictions from various object types

* Add docstrings for API config templates

* fix: Change the output - replace print with logging. Make adjustments to the imports.

* chore: rebase llm_trainer to tmp merging branch

* fix: add assembly to constant_repository

---------

Co-authored-by: ismonnar <ivanmonnar@gmail.com>
Co-authored-by: v1docq <ir@raxeltelematics.com>
Co-authored-by: Elarensba <153727503+Elarensba@users.noreply.github.com>
Co-authored-by: Irina Usoltseva <irinaslcv@gmail.com>
Co-authored-by: ismonnar <47676856+ismonnar@users.noreply.github.com>
Co-authored-by: Elarens <malyshevaveta@gmail.com>
Co-authored-by: leostre <strelllleo@mail.ru>
Co-authored-by: Андрей Глазков <glazkoff.andrej2014@yandex.ru>

Lopa10ko committed:
chore: add ts linguist repo infographic specification

Lopa10ko committed:
chore: delete mlruns/0 directory

chertila1 committed:
Add Sphinx docs generation from docstrings (#45)

Martakor1 committed:
upd: add DecomposerType(Enum)
fix: typing to Optional
fix: typing in Idecomposed
add decomposer_params
drop self.bias is not None checks
fix: test_layers.py - all test works now

Martakor1 committed:
fix: utils.hooks import changed
add: drop OperationParameters in fedcore.algorithm models

Martakor1 committed:
add __init__.py in external module

Martakor1 committed:
fix: add fields (rank, power, distortion_factor, random_init) for decomposer params in LowRankTemplate
fix imports

Martakor1 committed:
add: /llm_output (folder with model checkpoints) in gitignore

Martakor1 committed:
refactor: change decomposer_params. Now IDecomposed uses only initiated decomposer, instead of constructing it every time. LowRank creates decomposer in __init__. Methods moves to decomposer.py

Martakor1 committed:
fix: merge changes to low_rank_opt
additionally change __init__ of llm_trainer and BaseTrainer for compatibility
Also change BaseCompressionModel _init_model to correct places like _init_model_before_model_after
test_lr - works

Martakor1 committed:
fix: refactor BasePruner after merge, hooks works correctly, saving to registry marked as TODO
also clean rubbish doc
fix: some input_data.features troubles

Martakor1 committed:
fix: BaseQuantizer __init__ method and similar. Also drop many rubish docs, fix quantizer tests.
note: 2/3 quantizer tests doesnt work, but wrote that problem with device CPU and kernels???

Martakor1 committed:
fix: remove rubbish docs from hooks.py quantizer + fix import in test_llm_hooks.py

Martakor1 commented:
Коротко, что произошло: (коротко не получилось)
Я отребейзился на merge commit https://github.com/v1docq/FedCore/pull/42/changes
Комиты ребейза идут до upd: Add DecomposerType[Enum] [ab04f6c](https://github.com/v1docq/FedCore/pull/37/commits/ab04f6c6cf89a85e2cba396c75ec5f61651e6f80), после этого идут новые коммиты фиксы.
Что фиксилось глобально?
Учитывались старые, но изменённые _init в Quantizer, LowRank и Pruner и BaseNM и Base_Compr_Model. Постарался перенести логику с registry и выбором трейнера.
Теперь создание трейнера более универсально, но требуется проверка работы registry и идеи с копированием чекпоинтом и тд.
!Логика с registry видимо полетела, но ещё до меня, так как в исходной ветке (https://github.com/v1docq/FedCore/pull/33/changes?new_files_changed=true#top) было больше смысла , чем в мёрженой. Оставил комменты в TODO, так как я не могу отвалидировать пока эту логику.
Убрана излишняя документация
Починено много импортов хуков
Чуть-чуть переработаны IDecomposed модули и логика передачи в них decomposer упрощена
Поддержаны новые api_configs и LearningConfigTemplate
После мёржа поломались места с input_data.target - заменил на input_data.features.target во многих местах (Почему-то там теперь InputData а не CompressionInputData)
Убран лишний _get_example_batch в LowRank так как в бейз классе есть такой же
Исправлены тесты, теперь нет поддержки поля peft_strategy в  LearningConfigTemplate, peft_task передаётся в data_check вручную, не через конфиг.
Убраны импорты и оптимизирована логика работы с logger
Убран во второй раз (так как после ребейза он опять появился) finetune в pruner, так как с новыми хуками мы решили, что этой секции нет как отдельной, а регулируется через prune_each сколько steps будет до и после прунинга.

Протестировано, но нуждается в доработках (фиксирую состояние на этот момент):
Запускались базовые тесты low_rank, quantizer и pruner. Классы отрабатывают. Однако в quantizer теперь ошибка для static и qat, что-то не так с кернелами для CPU в методе summary() где quantizer.estimate_params(). Хотя раньше вроде работало.
2 теста отваливаются в transmla для reassembler (но это не касается хуков)
2 теста отваливаются в test_quant_utils, но они тоже на QDQWrapper
1 тест test_linear2 отваливается в test_layers - почему то RSVD декомпозер не может разложить матрицу W (какая-то ошибка с шейпами при matmul внутри)
Помимо этого, возможно осталось ещё много тестов, но не запущенных, так как всё ещё остаются ошибки с imports и кодом, который уже меньше касается хуков и базовых классов, а касается LLM экспериментов Иры.

Итог: авторам коммитов, попавших в мёрж нужно проверить свою логику. А именно: registry и сохранение моделей. LLM тесты и хуки на них. + желательно бы дофиксить тесты.

Дополнительные комментарии: 1) неопределённость InputData и CompressionData создаёт всё больше запутанностей в коде, как создала сейчас. 2) Избыточная автогенеренная документация выпилена/сокращена, но не везде, желательно довыпиливать, так как файлы раздуваются x2 в размере.

Предложения: лучше как можно скорее влить эту ветку, пусть и не полностью доработанную, иначе поддерживать разные версии кода хуков и прочего будет не реально, так как эта логика является довольно центральной в коде. Допилить тесты или что-то конкретное локально уже смогут авторы изменений.
